doc/mgmt/config_replace_rollback_HLD.md

```diff
@@ -0,0 +1,418 @@
+| time   						  | 30 .. 3600 seconds for which the configuration changes kept in running-configuration as trial stage |
+| label   						  | commit label string, upto 72 characters |
+| **Default**                                             | None |
+| **Command mode**                                        | Config-s |
+| **Roles**                                               | admin |
+| **Example**                                             | Sonic(config-s)#commit time 120<br>Sonic(config-s)#commit confirm |
+| **Support**                                             | 4.1 |
+Note: "commit" will be implemented by Broadcom. time and confirm options will be added to "commit" command
+### 3.2.3 REST API Support
+Regular RPC operation can be sent from REST. The server side implementation should be the same as exce mode CLI.
+
+### 3.2.4 gNMI Support
+
+
+# 4 Flow Diagrams
+# 4.1 Exec mode configuration replace command
+![Exec_mode_replace](config_replace_images/Exec_mode.png "Figure 1: Exec mode replace")
+	
+ __Figure 1: Exec mode configuration replace__
+
+# 4.2 Config session mode replace command
+![Config_session_mode_replace](config_replace_images/Config_session_replace.png "Figure 2: Config Session replace")
+	
+ __Figure 2: Config session mode replace__
+# 4.3 Config Session mode configuration commit timer command
+![Config_session_commit_timer](config_replace_images/Config_commit_time.png "Figure 3: Config Session mode commit time")  
+
+ __Figure 3: Config Session mode commit time__  
+Extend config session commit rpc for:   
+- Commit with no timeout value functions as default commit operation of configuration session.  
+- Handle commit confirm timer.  
+- On timer expiry,   
+	- Revert to the backup configuration. 
+	- Delete backup config file.
+	- Unmark trial stage.
+	- Close config session.  
+	  
+
+# 4.3 Config Session mode configuration commit confirm command
+![Config_session_commit_confirm](config_replace_images/Config_commit_confirm.png "Figure 4: Config Session mode commit confirm")  
+
+ __Figure 4: Config Session mode commit confirm__  
+Extend config session commit rpc for:  
+- Delete backup config file.  
+- Unmark trial stage.	
+- Close config session.  
+	
+
+# 5 Error Handling
+- Primary Encryption cannot be replaced using config replace key config-key password-encrypt and
+no key config-key password-encrypt commands trigger re-encryption of keys in backends and much more. 
+If Primary Encryption related commands are Created/Modified/Deleted through config replace,
+Error will be thrown
+	
+- User management commands cannot be replaced using config replace username and no username commands 
+directly configure local users on linux. If User management commands are Created/Modified/Deleted through config replace,
+Error will be thrown
+	
+- Break out of physical ports cannot be supported. If Break out related commands are Created/Modified/Deleted 
+through config replace, Error will be thrown 
+
```

Comment:
@kwangsuk (cc: @lisahnguyen @anand-kumar-subramanian @sachinholla @suresh-rupanagudi @rajendra-dendukuri @maheshwari-mayank )

What is the behavior if the DATABASE VERSION in the replace config_db.json file is not matching the current DATABASE VERSION on the switch? Will an error be thrown if this is detected?

This has bearing on upgrade/downgrade scenario for checkpoints in config-session.

`
"VERSIONS": {
    "DATABASE": {
        "VERSION": "version_4_1_1"
    }
}
`

Reply:
The config replace does not handle the config upgrade/downgrade functionality. It is intended to support the same version on both running and desired configuration. We check the version mismatch and enforce that the desired configuration must have the version attributes.